### ISSUE TITLE
Large-Scale Image Retrieval with Attentive Deep Local Features

### ISSUE THREAD

**Issue by chullhwan-song** [chullhwan-song/Reading-Paper#4]
https://arxiv.org/abs/1612.06321
https://research.googleblog.com/2018/03/google-landmarks-new-dataset-and.html?m=1
https://github.com/tensorflow/models/tree/master/research/delf

**Comment by chullhwan-song** [comment]
### abstract
* attention 기반 conv local descriptor
* google landmark 데이터셋 공개

### DELF (세미나 요약)
* SIFT와 매우 유사 > 딥으로하는 local descriptor라고 하면 됨.
* LIFT 같은 연구와는 비슷하지만 차이가 많이 존재
* keypoint 추출등의 모든 것을 하나의 network에서만 진행
* 테스트 단계에서만 multi-scale 처리
* keypoint 추출
   * fcn 이전의 conv feature map을 이용 - resnet 50 - **마지막 이전** conv layer 정도만. (실험에 의해)
      * resnet 50의 conv4_x로 시작되는 이름의 conv layer 이용
   * **attention network 이용**
     * 2-layer cnn : 1x1 filters
     * softplus(?) activation > 첨듣는말. [링크](https://papers.nips.cc/paper/1920-incorporating-second-order-functional-knowledge-for-better-option-pricing.pdf)
     * receptive field > grid 형태 so, sift keypint(x,y)와는 다름.
     * 이 단계에서 그냥 학습하면 그냥 attention network가 학습되어..자동적으로 keypoint가 감지된다.
* image pyramid 방식의 feature extraction
* fine-tuning - landmark 데이터셋을 이용, 140K, 586 종류의 landmark > triplet loss가 아닌 softmax(그냥
 이미지 분류)
* feature reduction 으로 PCA 사용
   * affine transformation 사용해도 괜찮을거 같다.- 저자가..
* rotation에 invariant하지 않다.

**Comment by chullhwan-song** [comment]
### 모델 구조
![image](https://user-images.githubusercontent.com/40360823/41751946-29eaee48-75ff-11e8-94e2-fc4252ade6cf.png)
* 그림의 왼쪽 학습 - imagenet 학습모델 기반(fine-tuning) softmax(classification loss) 학습 - 기존 우리가 알던 분류 학습
   * 데이터셋 : http://sites.skoltech.ru/compvision/projects/neuralcodes/
* 그림의 오른쪽 학습
   * adding attention network - 2 cnn : 노란색 layers
   * 빨간색은 Attention sore(알파)*conv_feature_map에 대해 weighted sum 한것으로 보임.
       * fcn network
   * 즉, 2개의 cnn network에 의해 spatial attention map이 형성되고 이를 기존 conv feature(노란색 layer아래 부분)와 곱을 취함. 이는 attention conv feature map의 형태가 되는 형태만 다름. 나머지는 같음.
       * 요 부분(spatial attention )이 learning에 의해 구해진다는 contribution이 존재
       * 그 담부터는 기존방법과 동일함 = conv > fcn > softmax)    
   * 파란색은 이것을 input 값으로 하여 softmax를 취함.   
* 결과적으로, 
   * 이 논문은 철저히 fine-grained(=landmark)성격의 학습을 통해 acc를 높혀놨음. 
   * 이를 위해, imagenet기반의 resnet50 pre-trained 모델을 가지고 landmark 데이터으로 학습(fig4, a)을 하고, 이를 다시, 위 fig4, b인 attention 을 학습하는것으로 보임(2번의 학습)
   * attention map은 그냥 local descriptor에 대한 feature selection 역할. 
* 그래서, 다음과 같이 철저히 landmark(건물만) attention이 가능햇던..
![image](https://user-images.githubusercontent.com/40360823/41751958-4c7cb2b6-75ff-11e8-9149-5ba85862bd74.png)
### indexing
* 이후, 빠르게 indexing하고 찾기 위해, PQ, K-Nearest 검색을 적용하고, 이후, 후보들에 대해 최종적으로 verification(RANSAC)을 함.
### trainset
* 위의 figure 4에서 a 와 b를 두번 finetuning하는건 overfitting이지 않나라고 예측하고 싶은데..잘된다니.ㅎ
* 왜 이런 생각이 드는 이유는 정리하자면 위에서 언급했듯이
   1. imagenet으로 학습된 pre-trained 모델을 이용함.
   2. LF : A. Babenko, A. Slesarev, A. Chigorin, and V. Lempitsky. Neural Codes for Image Retrieval. In Proc. ECCV, 2014 연구에서 사용된 landmark를 이용하여 finetuning -  figure 4에서 a
         * 3의 논문에서 일부 더 사용한듯~
         * [neuralcodes dataset](http://sites.skoltech.ru/compvision/projects/neuralcodes/) 
             * 586 landmark(클래스), 전체 140372
                * Oxf5k & Park6k 와 중복된 클래스는 제거 > 이 두개를 evalset으로 사용하는듯보임. 
                * 한클래스당 240개정도 되는듯
   3. LC : neuralcodes dataset을 SIFT 매칭기반으로 획득한 clean version
         * neuralcodes dataset은 noise가 존재하는가
         * 586 landmark(클래스), 35,382개 이미지 > LF와 같은 수의 클래스로 보아 같은 데이터셋이 확실한듯 
         * 클래스당 60개        
    * (Oxf5k/Par6k)은 landmark 데이터셋
    * 흠. 이런경우가..1바로 3으로 넘어가면 하면 성능이 좀 낮아지나??(시작을 landmark-> landmark로 가는게 더 좋을거같기도 한데 overfitting이 ...??)
* We use LF to train our attention model, and LC is employed to fine-tune the network for image retrieval.
   * 둘다 attention이면서, finetuning않거 아닌가? 더구나 LC할때는 같은 데이터를 두번째 쓰는것같은데..
* 자기네들 google landmark 데이터셋은 언제쓰이지? 학습할때 안쓰이나, kaggle에서는 train/val/test가 나위어져 있음(분석 필요)

### google landmark 데이터셋
* larger =  1,060,709 개 이미지
    * diverse landmarks = 12, 894
* query 이미지 = 111,036
* 실제 이개수는 맞지 않는듯보임. 증강되었나??
![image](https://user-images.githubusercontent.com/40360823/41752050-e0abc454-75ff-11e8-9050-6a440ab551bf.png)
![image](https://user-images.githubusercontent.com/40360823/41752054-e4b0c36a-75ff-11e8-99ee-26a4a3eb313a.png)

### 실험결과
![image](https://user-images.githubusercontent.com/40360823/41752058-eaf6ed58-75ff-11e8-9bb8-6c6489c3df1f.png)
